src/lib/index.js

```diff
@@ -459,10 +459,10 @@ lib.noneOrAll = function(containerIn, containerOut, attrList) {
  * @param {object} cd : calcdata trace
  * @param {string} cdAttr : calcdata key
  */
-lib.mergeArray = function(traceAttr, cd, cdAttr) {
+lib.mergeArray = function(traceAttr, cd, cdAttr, isNumber) {
     if(lib.isArrayOrTypedArray(traceAttr)) {
         var imax = Math.min(traceAttr.length, cd.length);
-        for(var i = 0; i < imax; i++) cd[i][cdAttr] = traceAttr[i];
+        for(var i = 0; i < imax; i++) cd[i][cdAttr] = isNumber ? +traceAttr[i] : traceAttr[i];
```

Comment:
What we should really do here is: drop this new `isNumber` argument and call `isNumeric`.

```js
for(var i = 0; i < imax; i++) {
  var v = traceAttr[i];
  cd[i][cdAttr] = isNumerix(v) ? +v : v;
}
```

This might result in a performance drop, but this sounds like an acceptable change until we drop `gd.calcdata` all together.

Reply:
What about keeping `isNumber` argument and only apply `isNumerix` on numeric fields e.g. `width` not `color`? That would help keep a better performance.

Reply:
Yeah we could do that. But remember `marker.color` can have numeric items. I presume that they are handling elsewhere (probably somewhere in `component/colorscale/`).

---

Moreover, passing a plain boolean to a function like 

```js
mergeArray(markerLine.width, cd, 'mlw', true);
```

is considered a bad pattern (to many) as you can't guess what the function does by looking at the line-of-code of the caller (here, you would need to open `src/lib/index.js` to see what the `true` does)

Here are two "better" patterns:

```js
mergeArray(markerLine.width, cd, 'mlw', {coerceNumber: true});

// or
mergeArray(markerLine.width, cd, 'mlw', function(v) { return isNumeric(v) ? +v : v; });
```

Personally I prefer the latter (as it is more general) and it consistent with how [`Lib.fillArray`](https://github.com/plotly/plotly.js/blob/master/src/lib/index.js#L479-L487) and [`Lib.castOption`](https://github.com/plotly/plotly.js/blob/218a9195832aab94b1d8bd0ca4653c159d2f25ce/src/lib/index.js#L498-L512) work.

Reply:
@etpinard thanks for the insight.
In https://github.com/plotly/plotly.js/pull/4056/commits/3ee7dbffe00cf5a899ee9858abe221e06e9e2e3f the second pattern you mentioned is applied to expand `mergeArray` functionalities.
Also in order to avoid requiring `fast-is-numeric` in `arrays_to_calcdata` steps in different traces, new functions were added to the `Lib`.

Reply:
Adding `Lib.mergeArrayCastPositive` is a very good idea. But less not add `Lib.mergeArrayCastNumber` and `Lib.mergeArrayCastIfNumber` if you're not going to use them anywhere.

Reply:
.... unless you're planning on fixing more `Lib.mergeArray` issues in this PR :smile:

Reply:
Let's do that in a separate PR.
For now I cut them from the Lib in https://github.com/plotly/plotly.js/pull/4056/commits/579a1b55c248350de6c2d461c87ac0a05bd0eef9.